README.txt
Minecraft Save Seed Reader

------------------------------

Check the GitHub wiki for more info!
https://github.com/thedarkfreak/Minecraft-Save-Seed-Reader/wiki

This program reads and displays the seed from all found minecraft Save files.   Just select a world's name from the dropdown list that shows up to get that world's seed.

If the save file is a Beta 1.8+ save file, you can switch the world between Creative and Survival mode by checking the box and hitting 'Save'.

If the save file is a Beta 1.9+ save file, you can toggle Hardcore mode on the save file on and off.

If you do not see your save files, or if you get an error stating the saves aren't found, try locating the Minecraft saves folder manually, either in the popup or by opening the Help menu and choosing "Select MC Save Location".  You have to select the 'saves' folder, not any of the world folders in the 'saves' folder.



The way the program works is as follows:


First, it determines what the default location of the .minecraft/saves folder would be, depending on what OS is found.

If it finds Windows, it uses the path %APPDATA%\.minecraft\saves\.

If it finds Mac OS X, it uses the path /users/<name>/Library/Application Support/.minecraft/saves/.

If it finds Linux, it uses ~/.minecraft/saves/.



Next, it looks into each sub-folder it finds in the .minecraft/saves/ folder, and looks for a level.dat file.  If it finds one, it assumes that this folder is a valid Minecraft save folder, and gets the LevelName tag from the level.dat file to use as the world's name.  If the level.dat file doesn't have a LevelName tag(older ones may not), it instead uses the folder's name as the world name.

When a world is selected from the dropdown list, it looks up the RandomSeed tag from that folder's level.dat file.

If you choose a save folder manually, it first checks the folder you choose for the existence of 'level.dat', 'session.lock', and the 'region' folder.  If it finds all three, it guesses that the user chose a world folder instead, and searches the parent folder for other saves.

CHANGELOG.txt
===============================================================
Version 1.7.3 - 
===============================================================

Refactoring.  

Got rid of the DirFilter private class.  Only used in one instance, replaced with an anonymous class.


===============================================================
Version 1.7.2 - 
===============================================================

Changed "Enable commands" to "enable cheats", as some commands are still accessible in single player.

Fixed issue #3, TAG_Int_Array is now implemented

===============================================================
Version 1.7.1 -
===============================================================

Changed the Creative mode checkbox to a combobox, to enable switching between Survival, Creative, and Adventure modes.

Changed the flow of logic in some places to prevent some code from executing twice.

===============================================================
Version 1.7 -
===============================================================

Added a toggle to enable/disable in-game single player commands, added in the 12w16a preview.

Added a workaround for an issue with JComboBox and duplicate item names(in this case, duplicate world names)

===============================================================
Version 1.6.2 -
===============================================================

Fixed issue #1, a bug preventing the overwrite confirmation screen from working properly.

Thanks, Evildropbear!

===============================================================
Version 1.6.1 -
===============================================================

Update to support the new tags added to the MC save file in Beta 1.9 PR 5.

If the tags are not handled properly, some Creative mode abilities can be 
enabled in Survival mode.

===============================================================
Version 1.6 -
===============================================================

Made the Seed textbox read-only, so people don't try to change it with the program.

Made the Save button disable after a press, so the world won't be saved again until
another change is made.

Updated the Tag class to allow editing of Tag values directly.

Added the ability to toggle hardcore mode on 1.9+ worlds.

Made the program more resilient against save corruption. Now, corrupted save files won't crash the program.

===============================================================
Version 1.5 -
===============================================================

Added the ability to toggle 1.8+ worlds between Creative and
Survival mode.

Also fixed a bug in the Tag class, that prevented TAG_Compound
entries from being modified correctly.

===============================================================
Version 1.4 -
===============================================================

Minor tweaks. Made the GUI use the look and feel of the current
system instead of Java's default look and feel, which makes it
look better no matter the platform you're on.

Also modified the 'About' menu text to include the github site
for the source code.

Changed behavior of folder selection dialog; now, it runs three checks after a folder has been chosen.  In the folder chosen, it checks for the existence of 'level.dat', 'session.lock', and the 'region' folder.  If all three exist, the folder chosen is most likely a world folder instead of the saves folder required, so the program automatically sets the chosen path to the parent folder.  Let me know if any problems come from this.

===============================================================
Version 1.3 -
===============================================================

Fixed a NullPointerException happening when searching for save files.  The program originally used the contents of the "LevelName" tag in the save file as the world's name in the dropdown list.  However, it turned out that older saves didn't have the "LevelName" tag, so trying to use it resulted in a NullPointerException.

As a fix, the program now checks if it can find a "LevelName" tag, and if it can't, it uses the name of the folder that the save file is in as the world name. (For example, if you have a save at .minecraft\saves\World1, which doesn't have a "LevelName" tag, it will use World1 as the world name.)

Also, added an error handling mechanism that will notify the user that an error has occured, and create an error log, named MinecraftSeed.error.log.  This will prevent it from failing silenty while it outputs the error information to a usually invisible console.  If it can't open the error log file, it will fall back to outputting to the console.

===============================================================
Version 1.2 -
===============================================================

Modified the GUI to use a dropdown list instead of a button.  The program will automatically search the default save directory for save files, and add them to the dropdown list.  Choose a save from the dropdown list to get that world's seed.

The directory it searches for saves in can be selected by the user by choosing Help -> Select MC Save Folder from the menu.

You will get a notice if the folder you select doesn't exist or if the folder you select doesn't appear to have any valid MC saves in it.  In that case, it will ask you to choose again.

===============================================================
Version 1.1 - 

src/minecraftSeed/MinecraftSeed.java
/*
This software and it's source are distributed under the terms of the Modified BSD License, detailed below.

Copyright (c) 2011-2012, Christopher Iverson
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of the Minecraft Save Seed Reader team nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
*/

package minecraftSeed;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.*;

import javax.swing.*;
import net.miginfocom.swing.MigLayout;


public class MinecraftSeed implements ActionListener {
	private Tag main;
	private JFrame frame;
	private JPanel panel;
	private JTextField text;
	private String savePath;
	private String[] filePaths;
	private String lastFolder;
	private int validNames;
	private JComboBox<Object> combo;
	private boolean hardcoreEnabled;
	private boolean commandsEnabled;
	private boolean abilitiesExist;     //To control tags added in Beta 1.9 PR 5.
										//If these tags are not changed properly, Creative mode
										//abilities can be enabled in Survival mode.
	
	private int gameType;
	private int playerGameType;
	private boolean usePGT;            //Stands for Use playerGameType.  Older saves didn't have this.
	private boolean initSave;          //This flag is enabled when a save file is first loaded, and before it's been interacted with.
									   //The main use for it is preventing the Gamemode combobox's action event from affecting data when first loaded.
	private JButton btnSave;
	private String selectedFilePath;
	
	//private final Integer version = MinecraftSeed.makeVersion(1,7,1);
	private final String version = "1.7.3";
	private JCheckBox cbHardcore;
	private JCheckBox cbCommands;
	private JLabel lblGamemode;
	private JComboBox<Object> cmbGamemode;
	
	public MinecraftSeed()
	{
		//Set the system look and feel
		
		try {
			UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
		} catch (ClassNotFoundException e) {
			e.printStackTrace();
		} catch (InstantiationException e) {
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			e.printStackTrace();
		} catch (UnsupportedLookAndFeelException e) {
			e.printStackTrace();
		}
		
		
		//Windows
		if(System.getProperty("os.name").toLowerCase().contains("windows"))
		{
			savePath = System.getenv("APPDATA") + "\\.minecraft\\saves\\";
		}
		
		//Mac
		if(System.getProperty("os.name").toLowerCase().contains("mac os x"))
		{
			savePath = System.getProperty("user.home") + "/Library/Application Support/minecraft/saves/";
		}
		
		//Linux(hopefully)
		if(System.getProperty("os.name").toLowerCase().contains("linux"))
		{
			savePath = System.getProperty("user.home") + "/.minecraft/saves/";
		}
		
		frame = new JFrame("Minecraft Save Seed Reader v" + version);
		frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		
		
		panel = new JPanel();
		
		combo = new JComboBox<Object>();
		combo.addActionListener(this);
		
		text = new JTextField();
		text.setColumns(20);
		text.setEditable(false);
		
		
		JMenuBar menuBar = new JMenuBar();
		JMenu helpMenu = new JMenu("Help");
		
		JMenuItem howToUse = new JMenuItem("How to use");
		howToUse.setActionCommand("use");
		howToUse.addActionListener(this);
		helpMenu.add(howToUse);
		
		JMenuItem newFolder = new JMenuItem("Select MC Save Folder");
		newFolder.setActionCommand("folder");
		newFolder.addActionListener(this);
		helpMenu.add(newFolder);
		
		helpMenu.addSeparator();
		
		JMenuItem about = new JMenuItem("About");
		about.setActionCommand("about");
		about.addActionListener(this);
		helpMenu.add(about);
		
		menuBar.add(helpMenu);
		frame.setJMenuBar(menuBar);
		panel.setLayout(new MigLayout("", "[28px][166px,grow][95px]", "[23px][][][]"));
		
		panel.add(combo, "cell 0 0,alignx left,aligny center");
		panel.add(text, "cell 1 0 2 1,growx");
		frame.getContentPane().add(panel);
		
		lblGamemode = new JLabel("Gamemode:");
		panel.add(lblGamemode, "cell 0 1,alignx trailing");
		
		btnSave = new JButton("Save Changes");
		btnSave.setEnabled(false);
		btnSave.setActionCommand("save");
		btnSave.addActionListener(this);
		
		cbCommands = new JCheckBox("Enable Cheats");
		cbCommands.setEnabled(false);
		cbCommands.setActionCommand("commandstoggle");
		cbCommands.addActionListener(this);
		
		cbHardcore = new JCheckBox("Hardcore Mode");
		cbHardcore.setEnabled(false);
		cbHardcore.setActionCommand("hardcoretoggle");
		cbHardcore.addActionListener(this);
		
		cmbGamemode = new JComboBox<Object>();
		cmbGamemode.setEnabled(false);
		cmbGamemode.setActionCommand("gamemode");
		cmbGamemode.addActionListener(this);
		panel.add(cmbGamemode, "cell 1 1 2 1,growx");
		panel.add(cbHardcore, "cell 1 2");
		panel.add(cbCommands, "cell 2 2");
		panel.add(btnSave, "cell 2 3");
		
		//Try to load the save data
		setupData();
		
		frame.pack();
		
		//Center on screen
		frame.setLocationRelativeTo(null);
		frame.setVisible(true);
	}
	
	/*private static int makeVersion(int major, int minor, int rev)
	{
		return (major << (3*8)) + (minor << (2*8)) + rev;
	}*/
	
	public static void main(String[] args) {

		try {
			new MinecraftSeed();
			
		} catch(Exception e) {  //If anything unexpected goes wrong in the main program,
			     				//write it to an error log
			try {
				FileOutputStream fos = new FileOutputStream("MinecraftSeed.error.log");
				e.printStackTrace(new PrintStream(fos));
				fos.close();
				
				JOptionPane.showMessageDialog(null, "An error occured." +
						"\nCheck the MinecraftSeed.error.log file for more information.", 
						"ERROR", JOptionPane.ERROR_MESSAGE);
				
			} catch (FileNotFoundException e1) {  //If the error log messes up,
												  //fall back to console
				e.printStackTrace();
			} catch (IOException e2) {
				e.printStackTrace();
			}
			e.printStackTrace();
		}
		
	}

	private void setupData()
	{
		boolean loop = false;
		
		//Check to reload data, used when checking folders with no valid save files
		do {
			loop = false;
			
			File file = new File(savePath);
			
			//If the save folder selected actually exists
			if(file.exists())
			{
				//Grab a list of sub-directories in the saves directory
				
				String [] tempPaths = file.list(new FilenameFilter() {
					public boolean accept(File dir, String name) {
						return new File(dir.getPath() + File.separator + name).isDirectory();
					}
				});
				filePaths = new String[tempPaths.length];
		
				validNames = 0;
				FileInputStream fis;
		
				combo.removeAllItems();
				for(String s : tempPaths)
				{
					//Check each sub-directory for a 'level.dat' file
					//If one is found, then assume it's a valid MC save folder
					file = new File(savePath + File.separator + s + File.separator + "level.dat");
					if(file.exists()) {
						filePaths[validNames] = file.getPath();
						
						lastFolder = file.getParent();
						
						try {
							fis = new FileInputStream(file);
							Tag temp = Tag.readFrom(fis);
							fis.close();
							
							//If the level.dat doesn't have a 'LevelName' entry,
							//go by the folder's name
							Tag name = temp.findTagByName("LevelName");
							if(name == null) {
								combo.addItem(makeObj(file.getParentFile().getName()));
							} else {
								combo.addItem(makeObj((String)name.getValue()));	
							}
							
						} catch (FileNotFoundException e) {
							e.printStackTrace();
						} catch (IOException e) {
							e.printStackTrace();
						} catch (Exception e)
						{
							//Unknown problem with loading save file.  Corrupted?
							JOptionPane.showMessageDialog(frame, "There was a problem reading the save file in " + lastFolder
									+ ".\n It will not be loaded in this program.", "Error", JOptionPane.ERROR_MESSAGE);
							continue;
						}
						
						validNames++;
					}
				}
				
				if(validNames == 0) 
			    { 
					//No valid save files were found
					JOptionPane.showMessageDialog(panel, "No valid save files were detected in current folder." +
							"\n\nPlease choose a new one.", "Error", JOptionPane.ERROR_MESSAGE);
					
					if(chooseSaveFolder()) loop = true;
					//This won't loop if the user cancels out of the folder selection dialog.
				}
			} else {  //Save folder doesn't exist
				JOptionPane.showMessageDialog(panel, "Save folder doesn't exist.\n\nPlease choose a new one.",
						"Error", JOptionPane.ERROR_MESSAGE);
				
				if(chooseSaveFolder()) loop = true;
				//Again, won't loop if the user cancels out of the folder selection dialog.
			}
		} while (loop);
	}
	
	private boolean chooseSaveFolder()
	{
		JFileChooser fc = new JFileChooser();
		fc.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
		
		int ret = fc.showDialog(panel, "Open Minecraft Saves Folder");
		if(ret == JFileChooser.APPROVE_OPTION) {
			savePath = fc.getSelectedFile().getPath();
			
			//Three tests to see if the folder chosen was a world folder, not the saves folder
			//Test one: check for existence of 'level.dat'
			if((new File(savePath + File.separator + "level.dat")).exists())
			{
				//'level.dat' exists, now check for 'session.lock'
				if((new File(savePath + File.separator + "session.lock")).exists())
				{
					//'session.lock' exists, now the final check: region subfolder.
					if((new File(savePath + File.separator + "region")).exists())
					{
						//At this point, we've almost indisputably gotten a world folder instead
						//of the saves folder.  Let's set 'savePath' to the parent directory.
						savePath = new File(savePath).getParent();
					}
				}
			}
			return true;
		}
		
		return false;
	}
	@Override
	public void actionPerformed(ActionEvent arg0) {
		//If a combobox item was selected
		if(arg0.getSource() == combo)
		{
			int val = combo.getSelectedIndex();
			
			//Make sure something was actually chosen
			if(val > -1)
			{
				//Disable the save button
				btnSave.setEnabled(false);
				
				try {
					selectedFilePath = filePaths[val];
					FileInputStream fis = new FileInputStream(new File(selectedFilePath));
					main = Tag.readFrom(fis);
					fis.close();
				} catch (FileNotFoundException e) {
					e.printStackTrace();
				} catch (IOException e) {
					e.printStackTrace();
				}
				
				usePGT = false;
				initSave = true;
				cmbGamemode.removeAllItems();
				cmbGamemode.addItem("Survival");
				
				
				text.setText(main.findTagByName("RandomSeed").getValue().toString());


				//Only if the save file has a 'GameType' entry will
				//we allow the user to switch between Creative on and off.
				Tag gametype = main.findTagByName("GameType");
				if(gametype==null)
				{
					cmbGamemode.setEnabled(false);
				}
				else
				{
					cmbGamemode.setEnabled(true);
					cmbGamemode.addItem("Creative");

					gameType = (Integer)gametype.getValue();
				}

				//Same as above, only with the "hardcore" toggle.
				Tag hardcore = main.findTagByName("hardcore");
				if(hardcore==null)
				{
					cbHardcore.setEnabled(false);
					cbHardcore.setSelected(false);
				}
				else
				{
					cbHardcore.setEnabled(true);

					//Check the box if hardcore is enabled
					hardcoreEnabled = ((Byte)hardcore.getValue() == 1);
					cbHardcore.setSelected(hardcoreEnabled);
				}

				Tag commands = main.findTagByName("allowCommands");
				if(commands==null)
				{
					cbCommands.setEnabled(false);
					cbCommands.setSelected(false);
				}
				else
				{
					cbCommands.setEnabled(true);

					commandsEnabled = ((Byte)commands.getValue() == 1);
					cbCommands.setSelected(commandsEnabled);
				}
				
				Tag player = main.findTagByName("Player");
				
				//The reason "Player" is also used here is to make sure we find the correct tag.
				//Player tag is separated to handle the case that a player tag doesn't exist in the save file.
				if(player != null)
				{
					Tag abilities = player.findTagByName("abilities");
					abilitiesExist = (abilities != null);

					if(abilitiesExist)
					{
						Tag mayBuild = abilities.findTagByName("mayBuild");
						if(mayBuild != null)
						{
							cmbGamemode.addItem("Adventure");	
						}
					}
					
					Tag playerGameMode = player.findTagByName("playerGameType");
					if(playerGameMode != null)
					{
						usePGT = true;
						playerGameType = (Integer)playerGameMode.getValue();
					}
				} 
				else  //if(player != null)
				{
					abilitiesExist = false;
					usePGT = false;
				}
				
				initSave = true;
				if(usePGT)
				{
					cmbGamemode.setSelectedIndex(playerGameType);
				}
				else
				{
					cmbGamemode.setSelectedIndex(gameType);
				}

			} // if (val > -1)
			
		} else { //One of the menu items was chosen, or the checkbox was checked,
			     //or "save" was pressed
			String cmd = arg0.getActionCommand();
			
			//"How to use" menu item
			if(cmd.equals("use"))
			{
				String message = "Choose the world's name from the dropdown list." +
								 "\nThe world's seed will show up in the textbox." + 
								 "\n\nIf the list is empty, click on Help > Select MC Save Folder," +
								 "\nand find and open the saves folder in the file chooser.\n" +
								 "\nIf the save is a Beta 1.8+ save, you can switch the world" +
								 "\nbetween Creative and Survival by checking the checkbox, then" +
								 "\nhitting Save.";
				
				JOptionPane.showMessageDialog(panel, message, "How to Use", JOptionPane.INFORMATION_MESSAGE);
			}
			
			//"About" menu item
			if(cmd.equals("about"))
			{
				String message = "Written by Chris Iverson. \n\n" +
								 "Source code available here:\n" +
								 "https://github.com/thedarkfreak/Minecraft-Save-Seed-Reader\n" +
								 "\nProblems? \nReport them either to the github site, in the forum topic,\n" +
								 "or send me an e-mail at cj.no.one@gmail.com.\n" +
								 "\nEnjoy!";
				
				JOptionPane.showMessageDialog(panel, message, "About Minecraft Save Seed Reader v" + version,
						JOptionPane.INFORMATION_MESSAGE);
			}
			
			//"Select MC Save Folder" menu item
			if(cmd.equals("folder"))
			{
				//Only load the data if the user selected a folder
				if(chooseSaveFolder()) setupData();
			}
			
			//Game mode changed
			if(cmd.equals("gamemode"))
			{
				if (cmbGamemode.getSelectedIndex() > -1) 
				{
					
					if (!initSave) {
						gameType = cmbGamemode.getSelectedIndex();
						Tag gamemode = main.findTagByName("GameType");
						gamemode.setValue(gameType);
						
						if (usePGT) {
							playerGameType = gameType;
							Tag pgt = main.findTagByName("Player").findTagByName("playerGameType");
							pgt.setValue(playerGameType);
						}
						//Enable/disable all special abilities depending on selection.
						//while "Flying" is only used by the game to indicate whether or not the player is currently flying,
						//it is toggled here to make sure a player can't remain flying when Creative is turned off.
						if (abilitiesExist) {
							Tag abilities = main.findTagByName("Player")
									.findTagByName("abilities");

							Tag temp;

							byte enable = (byte) ((gameType == 1) ? 1 : 0);

							temp = abilities.findTagByName("instabuild");
							temp.setValue(enable);

							temp = abilities.findTagByName("invulnerable");
							temp.setValue(enable);

							temp = abilities.findTagByName("mayfly");
							temp.setValue(enable);

							temp = abilities.findTagByName("flying");
							temp.setValue(enable);

							enable = (byte) ((gameType == 2) ? 1 : 0);

							temp = abilities.findTagByName("mayBuild");
							if (temp != null)
								temp.setValue(enable);
						}
						//Enable the "Save" button
						btnSave.setEnabled(true);
					} //if(!initSave)
					else
					{
						initSave = false;
					}
				}
			}
			
			
			//Hardcore mode toggled
			if(cmd.equals("hardcoretoggle"))
			{
				hardcoreEnabled = !hardcoreEnabled;
				
				Tag hardcore = main.findTagByName("hardcore");
				hardcore.setValue((byte)(hardcoreEnabled? 1 : 0));
				
				//Enable the "save" button
				btnSave.setEnabled(true);
			}
			
			if(cmd.equals("commandstoggle"))
			{
				commandsEnabled = !commandsEnabled;
				
				Tag commands = main.findTagByName("allowCommands");
				commands.setValue((byte)(commandsEnabled? 1 : 0));
				
				btnSave.setEnabled(true);
			}
			
			//Save button pressed
			if(cmd.equals("save"))
			{
				int chosen = JOptionPane.showConfirmDialog(frame, "Are you sure you want to overwrite your save file?", 
						"Overwrite", JOptionPane.YES_NO_OPTION);
				
				if(chosen == JOptionPane.YES_OPTION)
				{
					try {
						FileOutputStream fos = new FileOutputStream(new File(selectedFilePath));
						main.writeTo(fos);
						fos.close();
					} catch (FileNotFoundException e) {
						// TODO Auto-generated catch block
						JOptionPane.showMessageDialog(null, "Problem saving file: File Not Found", "File Not Found", JOptionPane.ERROR_MESSAGE);
					} catch (IOException e) {
						// TODO Auto-generated catch block
						JOptionPane.showMessageDialog(null, "Problem saving file: IO Error", "IO Error", JOptionPane.ERROR_MESSAGE);
					}
					
					JOptionPane.showMessageDialog(frame, "File saved!");
					
					btnSave.setEnabled(false);
				}
			}
		} // else
		
	} //actionPerformed
	
	private Object makeObj(final String item)  {
		return new Object() { public String toString() { return item; } };
	}


}
